### PR TITLE
Reduce goroutine leakage in test/integration/controlplane/transformation

### LINKED ISSUE #111674: reduce goroutine leakage in `test/integration/controlplane/transformation`
Following up on https://github.com/kubernetes/kubernetes/pull/111126#discussion_r935087747 to shutdown the gRPC goroutines and reduce the number of leaked goroutines.

/sig auth

### PR DESCRIPTION
**Pull request by sanwishe**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
fix leaking goroutines in `test/integration/controlplane/transformation`

#### Which issue(s) this PR fixes:

Fixes #111674

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by enj** [comment]
Instead of a manual stop method that is a no-op in many cases I think it would be better to have a context be used to cancel any implementations that explicitly need it.

**Review by yangjunmyfm192085 — COMMENTED** [review]
/ok-to-test

File: test/integration/framework/etcd.go [hunk 1]
```diff
@@ -198,7 +198,7 @@ func EtcdMain(tests func() int) {
 		// like k8s.io/klog/v2.(*loggingT).flushDaemon()
 		// TODO(#108483): Reduce this number once we address the
```
Comment (yangjunmyfm192085) [inline test/integration/framework/etcd.go:199]: ToDo can be deleted.

**Comment by aojea** [comment]
> Instead of a manual stop method that is a no-op in many cases I think it would be better to have a context be used to cancel any implementations that explicitly need it.

@enj this is coming from https://github.com/kubernetes/kubernetes/issues/108483, it is not about context cancellation it is about storage shutdown, it needs a function that allows to stop and destroy all resources

File: staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go [hunk 2]
```diff
@@ -100,6 +106,10 @@ func (t *MutableTransformer) TransformToStorage(ctx context.Context, data []byte
 	return transformer.TransformToStorage(ctx, data, dataCtx)
 }
 
+func (t *MutableTransformer) Stop() error {
+	return t.transformer.Stop()
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:110]: the other operations are protected by a lock, don't we need to use the lock here too?

File: test/integration/framework/etcd.go (hunk 1 above)
Comment (aojea) [reply test/integration/framework/etcd.go:199]: hopefully can be deleted soon, but there are still some other leaks https://github.com/kubernetes/kubernetes/issues/108483#issuecomment-1195714805

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 3]
```diff
@@ -348,21 +348,23 @@ func newETCD3Storage(c storagebackend.ConfigForResource, newFunc func() runtime.
...
+		transformer = value.IdentityTransformer
+	}
+
 	var once sync.Once
 	destroyFunc := func() {
 		// we know that storage destroy funcs are called multiple times (due to reuse in subresources).
 		// Hence, we only destroy once.
 		// TODO: fix duplicated storage destroy calls higher level
 		once.Do(func() {
+			transformer.Stop()
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:362]: does order matters?

**Comment by aojea** [comment]
/assign @wojtek-t 

he is more familiar with this code

File: staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go (hunk 2 above)
Comment (sanwishe) [reply staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:112]: sorry for missed this, already updated.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 3 above)
Comment (sanwishe) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: It matters, and i have move it after `client.close()`

**Comment by enj** [comment]
> > Instead of a manual stop method that is a no-op in many cases I think it would be better to have a context be used to cancel any implementations that explicitly need it.
> 
> @enj this is coming from #108483, it is not about context cancellation it is about storage shutdown, it needs a function that allows to stop and destroy all resources

I understand what this code is trying to do.  Contexts can be used to control lifetime, in the same way a manual stop method can.  I do not want to expand the storage transformer interface with a method that one can forget to call.  Instead, any implementation that needs any form of cleanup should take a context in its constructor and automatically perform said cleanup when the context is canceled.

**Comment by 249043822** [comment]
a kms transformer which uses a grpc-service but doen't provide a mechnism to close the connection of grpc-service , so we should assure connection close while apiserver shutdown.  since the lifecycle of the transformer is consistent with apiserver, does listenning a shutdown signal(or a stopCh) for grpc-service closing more suitable than a context?  @aojea @enj @liggitt

**Comment by wojtek-t** [comment]
> a kms transformer which uses a grpc-service but doen't provide a mechnism to close the connection of grpc-service , so we should assure connection close while apiserver shutdown. since the lifecycle of the transformer is consistent with apiserver, does listenning a shutdown signal(or a stopCh) for grpc-service closing more suitable than a context? @aojea @enj @liggitt

Storage is being destroyed at the very end:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L464
We still don't have graceful shutdown of watches (which can use transformers), so we need to ensure  that transformers won't get shutdown too early.

> Instead, any implementation that needs any form of cleanup should take a context in its constructor and automatically perform said cleanup when the context is canceled.

@enj - I don't have very strong opinion on that and if we can make that work with contexts - I'm fine with it.
That said, currently transformers (in particular the KMS ones) are created very early, before even the apiserver config is created:
https://github.com/kubernetes/kubernetes/blob/18ce801beac34489f313d6b1223a1a5f07200cb6/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L245
[it's called at the level of initializing options]
So this approach would require significant changes to be implemented.

OTOH, the current Destroy/Stop-like approach is relatively easy - we don't expect people to be doing it on their own. It's the responsibility of the server to do that on shutdown, and as such I don't think that the concern of someone forgetting to call it is a valid one.

**Comment by enj** [comment]
We already have existing bugs for CRDs not supporting encryption at rest because the wiring is incorrect so I don't buy the argument that "we won't forget because we only need to do it in one place in the API server" - nothing enforces that invariant.  If we want to get the go routine number down before test freeze with this change, that is fine.  I have to refactor a lot of this code for KMS v2 so I can clean it up then.

**Comment by wojtek-t** [comment]
> We already have existing bugs for CRDs not supporting encryption at rest because the wiring is incorrect so I don't buy the argument that "we won't forget because we only need to do it in one place in the API server" 

My point was that cost is not per transformer-author. We have a fixed cost to properly wire the Initialize/Destroy interface (and yes, I agree with you that it's more than one place, because e.g. we do that differently for CRDs vs built-ins).
But I disagree that the model with destroy is more complex - this is roughly symmetric. In one case (your proposal) we need to enforce proper initialization (also not exactly the same across CRDs and built-ins), in this case, we need to ensure proper deletion. I don't agree that one is visibly simpler than the other.

I don't know what your plans for KMSv2 are exactly and I admit that the additional interface method isn't perfect. But it has one advantage of forcing people to think about the lifecycle, which definitely isn't the case for many authors now (not specific to transformers - it's the same for many controllers, plugins, etc.).

I don't think it's super critical for 1.25, so if you are planning to changing that soon (in 1.26) I'm fine with waiting to see if we can make that work in a simpler way then. If this will take multiple releases to get there, I think we should try merging something like this (I didn't do detailed review of it - just skimmed it).

**Comment by enj** [comment]
KMS v2 is part of my day job so it will definitely get done 😉

@liggitt had expressed wanting this go routine bit fixed in 1.25 so this may be fine for now (I haven't looked thoroughly).

**Comment by aojea** [comment]
> KMS v2 is part of my day job so it will definitely get done 😉

nice

@wojtek-t can you take a look and judge if we should add it to 1.25, we are still on the testing freeze period https://www.kubernetes.dev/resources/release/#summary

**Comment by wojtek-t** [comment]
> @wojtek-t can you take a look and judge if we should add it to 1.25, we are still on the testing freeze period

I took a deeper look into this PR and this LGTM.
I think that even if longer-term we decide to change the interface (as part of changes for KMSv2) and get rid of the `Stop()` method this is fine.
But given I don't know the exact plans for KMSv2 I would like to hear an ACK from @enj before approving.

From purely test-freeze perpective, that looks reasonable to merge.

**Comment by enj** [comment]
I have to do a lot of refactors to the KMS code to support the v2 feature set so this will not be a big issue for me to update.  @wojtek-t you will have to approve the API server code changes.

/lgtm
/approve

**Comment by enj** [comment]
Test freeze for 1.25 was `18:00 PDT Tuesday 9th August 2022` so I think we have missed the window for 1.25...

**Comment by enj** [comment]
/lgtm cancel

(since we missed the deadline)

**Comment by enj** [comment]
I have opened #111986 to fix this bug without expanding the transformer interface.

**Comment by enj** [comment]
@sanwishe thanks for your contribution.

This issue was resolved in #111986

/close

**Pull request closed without merging**